### ISSUE TITLE
Permanent download dir

### ISSUE THREAD

**Issue by NS-Clone** [Feodor2/Mypal68#547]
мало того что в браузере есть такое лгбт как browser.download.lastDir.savePerSite;false
вот еще один косяк на подобии хромов

вроде как для save as
вместо browser.download.dir
используетсо browser.download.lastDir

в итоге если разово сделать save as в другую папку
все save as пойдут туда же навсегда
конечно для 1-й сессии это может быть и не так плохо а может иногда и полезно
но вот при запуске браузера
эту бы муть обнулять до дефолтного browser.download.dir

**Comment by zanud** [comment]
> в итоге если разово сделать save as в другую папку
> все save as пойдут туда же навсегда

Разве? У меня в настройках стоит "Всегда выдавать запрос на сохранение файлов", и перед сохранением каждого файла выскакивает окошко выбора - куда сохранять и под каким именем.

**Comment by NS-Clone** [comment]
> и перед сохранением каждого файла выскакивает окошко выбора - куда сохранять и под каким именем.

Угу и каждый раз по три раза добавлять к имени какое нить sdjfvnsd чтоб только сохранить
хромой хотябы подставляет (1...100500) если такой файл уже в папке есть
а куча быдлo сайтеков щитает своим долгом выдавать разные файлы с одинаковым именем....

**Comment by zanud** [comment]
> Угу и каждый раз по три раза добавлять к имени какое нить sdjfvnsd чтоб только сохранить
> хромой хотябы подставляет (1...100500) если такой файл уже в папке есть
> а куча быдлo сайтеков щитает своим долгом выдавать разные файлы под одинаковым именем....

С одной стороны, случай, когда в каталоге уже есть файл с таким именем, к теме жалобы отношения не имеет.

А с другой, автоматическое переименование путём добавления номера - тоже хреновое решение в такой ситуации. И я безусловно предпочитаю руками задавать новые осмысленные имена файлов, чтобы не гадать потом, какой номерной файл откуда взялся.

**Comment by NS-Clone** [comment]
> > Угу и каждый раз по три раза добавлять к имени какое нить sdjfvnsd чтоб только сохранить
> > хромой хотябы подставляет (1...100500) если такой файл уже в папке есть
> > а куча быдлo сайтеков щитает своим долгом выдавать разные файлы под одинаковым именем....
> 
> С одной стороны, случай, когда в каталоге уже есть файл с таким именем, к теме жалобы отношения не имеет.
> 
> А с другой, автоматическое переименование путём добавления номера - тоже хреновое решение в такой ситуации. И я безусловно предпочитаю руками задавать новые осмысленные имена файлов, чтобы не гадать потом, какой номерной файл откуда взялся и что содержит.

когда тте надо натыкивать сотни таких имен
ты задолбаешся вводить им осмысленные имена

**Comment by zanud** [comment]
> когда тте надо натыкивать сотни таких имен
> ты задолбаешся вводить им осмысленные имена

У меня только однажды была ситуация, когда таких файлов нужно было сохранить много - штук 50. И да, я добросовестно приписывал в конце осмысленные хвостики. Потому что: а как иначе? Там же сайт не только разные архивы с одним и тем же именем отдавал, но и внутри архивов имена подкаталогов и файлов были одинаковые. Вот если бы я не руками что-то писал, а доверил это дело автоматической нумерации, то как бы я потом разобрался, в каком архиве что?
Сайт, кстати, назывался GitHub 🤣 

> или дружно переименовать

Руками переименовывать? Или как?

**Comment by NS-Clone** [comment]
> Руками переименовывать? Или как?

например advanced renamer-ом
чтоб проставить нормальные номерки сзади

**Comment by zanud** [comment]
Ну, будет у меня 500 архивов Document000.zip - Document149.zip, внутри каждого из которых лежит Document.doc. Как мне нужный документ там найти-то, когда он понадобится?

**Comment by NS-Clone** [comment]
> Ну, будет у меня 150 архивов Document000.zip - Document149.zip, внутри каждого из которых лежит Document.doc. Как мне нужный документ там найти-то, когда он понадобится?

если они все будут свелены в datasheets/ymf262
особых проблем с пониманием что в них не будет

**Comment by zanud** [comment]
> если они все будут свалены в datasheets/ymf262
> особых проблем с пониманием что в них не будет

Да, с пониманием, что в них, не будет. Вопрос был: как при необходимости найти в этой куче архивов нужный документ? Распаковывать и просматривать все по очереди?

А если ещё и само содержимое тех файлов не позволяет установить, тот это файл или нет? Например, там просто безымянные таблицы чисел. Или там разные версии одного и того же длинного текста, отличающиеся лишь несколькими словами где-то в середине.

**Comment by Feodor2** [comment]
Я не понимаю что вы хотите, всё с этим связанное работает как надо.
Если спрашивать то спрашивать, нет то нет, в случае неспрашивать можно задавать куда будет писать.

**Comment by NS-Clone** [comment]
> Я не понимаю что вы хотите, всё с этим связанное работает как надо. Если спрашивать то спрашивать, нет то нет, в случае неспрашивать можно задавать куда будет писать.

оно НЕ РАБОТАЕТ КАК НАДО
есть ПАПКА ДЛЯ ЗАГРУЗОК
файлы грузятсо хрен знает куда но НЕ В НЕЁ

**Comment by zanud** [comment]
> то как добавить в about:config что либо при запуске?
> дайте я сам прибью то что мне надо гвоздями
> и чтоб оно не отпадало на каждый пчих

Всю жизнь это делалось через самодельный файл `user.js` в профиле. В него можно помещать всё то же самое, что и в `prefs.js`, однако кроме функции `user_pref()` можно использовать ещё и 
несколько других, позволяющих прибивать более надёжно - не только гвоздями, но и шурупами.
Вот те ссылка: https://udn.realityripple.com/docs/Mozilla/Preferences/A_brief_guide_to_Mozilla_preferences

> у меня включено не спрашивать
> а сохранять в конкретную папку

Как же ты в таком случае сумел получить запрос на сохранение? :
> после того как из save as выбрал РАЗОВО другой путь для сохранения

И, кстати, проверь у себя в Основных настройках блок "Приложения" - там для конкретных типов файлов вполне могут оказаться прописанным "Сохранить файл", а эта настройка умеет запоминать и каталог, куда сохранять, и для файлов этого типа использоваться будет именно этот каталог, а не общий.

**Comment by NS-Clone** [comment]
> И, кстати, проверь у себя в Основных настройках блок "Приложения"

ух еще ж одни грабли
надо будет потыкать

**Comment by Feodor2** [comment]
То есть ты хочешь чтобы "сохранить как" не запоминало последнее место, это можно, но таки опять, а кто ещё так тоже хочет? Опять можеш найти почеу мозила это сделала.

**Comment by NS-Clone** [comment]
> То есть ты хочешь чтобы "сохранить как" не запоминало последнее место

ну типо до
а лучшо чтоб оно обнулялось при старте
иначе на второй, третий день
файлы идут потом неизвестно куда

**Comment by zanud** [comment]
> То есть ты хочешь чтобы "сохранить как" не запоминало последнее место

Нет, он этого не хочет!!! У него в настройках выбрано автоматическое сохранение в определённый каталог, и поэтому окно "Сохранить как" у него не выскакивает. Но при этом, судя по жалобе, у него не все файлы сохраняются в тот каталог; некоторые - куда-то в другое место.

Это кажется невозможным, но мне приходилось руками ковырять файл mimeTypes.rdf (визуализируемый блоком "Приложения" в "Настройках"), и я знаю, какие чудеса там возможны.

**Comment by NS-Clone** [comment]
> и поэтому окно "Сохранить как" у него не выскакивает.

выскакивает и с полным вводом имени
и с простым save cancel (и вообще мне этот вариант удобен)
и вообще иногда сохраняет просто так молча и сразу

вощем там под многими типами файлов always ask
(взаимоисключащие параграфы :rofl:)

**Comment by zanud** [comment]
> видимо оно переопределяет всегда сохранять

Ну вот, видишь, и браузер калечить не требуется.

**Comment by NS-Clone** [comment]
> > видимо оно переопределяет всегда сохранять
> 
> Ну вот, видишь, и браузер калечить не требуется.

на оборот он щас покалеченный
5 разных настроек
но все равно сайтек лучше знает что тебе надо

**Comment by zanud** [comment]
> но все равно сайтек лучше знает что тебе надо

Это не сайт. Это браузер. Каждый получаемый с любого сайта файл сопровождается каким-то MIME-типом. Когда Mypal видит некий тип в первый раз, он записывает его в упомянутый мной выше файл. А потом мы видим это файл через блок "Приложения". И даже редактировать немного можем.

> а папка загрузков должна быть ОДНА

Так она одна и есть. Но - для тех MIME-типов, с которыми браузер "познакомился" после того, как ты задал в настройках автоматическое сохранение. А те типы, которые попали в файл раньше, так и остались в нём со старыми значениями - их теперь руками править надо.

> свою для каждого сайта я выключил

Но в mimeTypes.rdf для некоторых типов остались заданными свои каталоги.

**Comment by NS-Clone** [comment]
> с любого сайта файл сопровождается каким-то MIME-типом

я так понел что там кроме расширений может быть еще какаято муть

**Comment by zanud** [comment]
Да, конечно. Ты что, ни разу не сталкивался с термином "MIME-тип"?
Эта "муть" для браузера является основной, ему как раз на расширения файлов смотреть не положено.

А MIME-типы в ом окошке как раз и прописаны. Просто иногда они там видны в явном виде (два идентификатора латиницей через косую), а иногда в "человеческом" - и тогда сам тип можно увидеть во всплывающей возле курсора подсказке в скобках.
Иногда, правда, в этой подсказке типа не оказывается, но таких строк не много, и наверное они от тех редких ситуаций, когда сайт прислал файл без указания типа - вот тогда браузеру и приходится ориентироваться на расширение.

**Comment by zanud** [comment]
> Да, конечно. Ты что, ни разу не сталкивался с термином "MIME-тип"? Эта "муть" для браузера является основной, ему как раз на расширения файлов смотреть не положено.
> 
> А MIME-типы в том окошке как раз и прописаны. Просто иногда они там видны в явном виде (два идентификатора латиницей через косую), а иногда в "человеческом" - и тогда сам тип можно увидеть во всплывающей возле курсора подсказке в скобках. Иногда, правда, в этой подсказке типа не оказывается, но таких строк не много, и наверное они от тех редких ситуаций, когда сайт прислал файл без указания типа - вот тогда браузеру и приходится ориентироваться на расширение.

**Comment by NS-Clone** [comment]
> ему как раз на расширения файлов смотреть не положено.

отож
потом тте 30% сайтеков говорит что нужно архив не качать
а открывать
а браузер молча выбирает open as по дефолту
а кто то там крякает про неположено неположено

**Comment by zanud** [comment]
> потом тте 30% сайтеков говорит что нужно архив не качать
> а открывать

Сайт не может сказать браузеру такое. Протоколом HTTP это не предусмотрено. Всё определяется настройками браузера.

У тебя же сейчас автоматическое сохранение в настройках задано? Ну вот и удали файл mimeTypes.rdf из профиля.

Mypal его потом создаст и начнёт заново заполнять. А значением по умолчанию для каждого нового типа будет то, что стоит в общих настройках (т.е. "Сохранять автоматически в заданный каталог").

**Comment by NS-Clone** [comment]
@Feodor2 вот только что
взял PNG там было always ask
и по дефолту было выбрано save as
переключил на save file
стало сохранятсо молча
переключил на always as
и по дефолту теперь выбираетсо open as

**Comment by zanud** [comment]
> ну и вощем судя по всему мне нужен именно always ask (ибо на нем видно что сохранение действительно пошло) + сброс save as папки в дефолтное положение

Создай в профиле файл `user.js` и скопируй в него из `prefs.js` строчки `user_pref("browser.download.useDownloadDir", false);` и `user_pref("browser.download.lastDir", "_что_там_у_тебя_");`

Всё, получишь именно желаемое: перед сохранением каждого файла будет окно с возможностью выбрать каталог/имя, и после каждого перезапуска браузера каталог сохранения будет тот, который прописан в файле `user.js`.

_Только я чего-то не понимаю: ты же писал, что хочешь, чтобы файлы сохранялись автоматически, без окна выбора каталога._

> a я всегда думол что always ask это выдавать полную менюшку с возможнсотью переименовать и выбрать куда сохранять как раз очень подходит под это названиё

Да, именно это always ask и означает.

> а всегда сохранять перебивалось настройками недоступными из конфига? о_О а я думол что оно так и работаит :)

Мы с тобой это уже когда-то копали. Если включено always ask, то перед сохранением файла показывается окно с выбором: "Сохранить файл", "Открыть в..." и галочкой "Запомнить выбор".
Если выбираешь "Сохранить файл", то показывается второе окно - с выбором каталога.
От первого окна можно избавиться, установив `browser.download.forbid_open_with = true`.

**Comment by Feodor2** [comment]
В "сохранить как" всегда выбирается куда и оно запоминает последнее, как я понял дела в этом, или нет?

**Comment by zanud** [comment]
> как я понял дела в этом, или нет?

Не совсем. В части _«В "сохранить как" всегда выбирается куда и оно запоминает последнее»_ - всё так и есть, и оно нормально работает, и трогать там ничего не надо.

Но некоторые несознательные личности своими несознательными манипуляциями могут добиться того, что для некоторых MIME-типов файлов в `mimeTypes.rdf` окажутся прописанными другие каталоги (и даже другой режим скачивания). И тогда по умолчанию для этих типов будет использоваться не общий каталог из `browser.download.lastDir`, а их индивидуальный.

Это не проблема программы, а исключительно дело рук самого пользователя. И лечится оно тоже исключительно пользовательскими мерами (я выше указал варианты).

**Comment by NS-Clone** [comment]
> Это не проблема программы, а исключительно дело рук самого пользователя

во какой умный тут нашелсо

**Comment by zanud** [comment]
> проблема что оно запоминает не для сессии
> а навсегда

`user_pref()` в `user.js` как раз и обеспечивает запоминание ровно на одну сесссию

**Comment by NS-Clone** [comment]
> что для некоторых MIME-типов файлов

еще интересней
вот основной браузер
сайтек вкоторый хожу каждый день
расширение файла которое качаю каждый
никакие ассоциации в реестре 3 года не менял
и ВНЕЗАПНО оно мне начало предлогать открыть как...

**Comment by NS-Clone** [comment]
и на тему 2-х папок
вот для картинок ТОЛЬКО save as
и даже если открыл ее во вкладке
у тебя все равно менюшку с полным выбором пути (подозреваю что это опять же lastdir)

**Comment by zanud** [comment]
> вот для картинок ТОЛЬКО save as
> и даже если открыл ее во вкладке
> у тебя все равно выкидывает менюшку с полным выбором пути (подозреваю что это опять же lastdir)

Те типы, которые Firefox/Mypal умеет показывать сам (HTML, GIF, JPG, PNG, ...) не подчиняются общим правилам и никогда не попадают в `mimeTypes.rdf`.

Собственно, все мои познания в структуре и возможностях `mimeTypes.rdf` происходят именно от того, что однажды мне понадобилось сделать так, чтобы JPG-и сохранялись в заданный каталог без вопросов. И тоге я сумел-таки вписать `image/jpeg` в `mimeTypes.rdf` руками.

**Comment by NS-Clone** [comment]
> не подчиняются общим правилам и никогда не попадают

у меня вполне в этом списке jpg и png
(ну в самом виевере в настройках)

**Comment by NS-Clone** [comment]
> Те типы, которые Firefox/Mypal умеет показывать сам (HTML, GIF, JPG, PNG, ...), не подчиняются общим правилам и никогда не попадают в `mimeTypes.rdf`

умя вполне jpg и png в этом списке
(тот который в настройках)

**Comment by zanud** [comment]
А у меня ничего из мной перечисленного там нет, несмотря даже на то, что со времён, когда я Firefox ковырял, кое-что изменилось, и теперь Mypal покезывает там не только то, с чем сам сталкивался, но и то, что в реестре Windows прописано.

Картинку с этим окном показать можешь?

**Comment by NS-Clone** [comment]
> Картинку с этим окном показать можешь?

огрызок
![2024-12-30_045440](https://github.com/user-attachments/assets/c916cfe8-cdf4-4621-9ef2-a942bf2035ce)

там жо жипег, гиф, вебп

> что в реестре Windows прописано.

как я уже говорил
при этом у нескольких одинаковых портативных ббраузеров
разное поведение

хотя вот название расширения
явно подтянуто из реестра

**Comment by NS-Clone** [comment]
> Если браузер сделан портативным какими-то внешними средствами

фаирфокс и так портативен из коробки

**Comment by zanud** [comment]
> фаирфокс и так портативен из коробки

Что не мешает выпускать и его версии, сделанные с привлечением всяких левых средств.
Причём и сами мозилловцы тоже так поступают: http://portableapps.com/go/FirefoxPortableESR

> какой еще кусок реестра?

Любой, какой захочется  авторам сборки. Ты что, не знаешь, по какому принципу все такие портативные сборки работают?

**Comment by NS-Clone** [comment]
на тему open as по дефолту
добавил нужный формат в handlers.json
и оно заработало с save as по дефолту
но там опять нада телепатия
чтоб угадать какое там mimetype какойто мудaк придумал для нужного формата файла

**Comment by NS-Clone** [comment]
а на тему хотелок про папку загрузок
понаблюдав еще
наверно хотелось бы папку загрузок ЖЕЛЕЗОБЕТОННО ПРИБИТО к папке загрузок
save as в другую папку это ЕДИНИЧНЫЙ СЛУЧАЙ
потом тебе для следующего файла нужно шагать и искать старую папку загрузок
что вощем то ППЦ

**Comment by zanud** [comment]
> наверно хотелось бы папку загрузок ЖЕЛЕЗОБЕТОННО ПРИБИТО к папке загрузок
> save as в другую папку это ЕДИНИЧНЫЙ СЛУЧАЙ

Не очень понятно, чего тебе хочется.

Когда в настройках браузера выбираешь сохранение файлов без вопросов, то там же задаёшь и каталог, куда будут сохраняться все файлы.

Если даже напорешься на исключение и для какого-то файла запрос выскочит, то это никак не повлияет на тот заданный каталог, и все последующие автоматические сохранения всё равно будут идти в него.

Если же ты в настройках выбираешь сохранение файлов с запросом имени и места, то вполне естественно, что браузер каждый раз будет запоминать место, куда пользователь выбрал сохранить файл, и в следующем запросе первым делом предложит его же.

> потом тебе для следующего файла нужно шагать и искать старую папку загрузок
> что вощем то ППЦ

Да, неудобно.

Но если считаешь, что каждый раз открывать диалог сохранения на одном и том же каталоге это удобно, то используй файл user.js, зафиксировав там `browser.download.lastDir` подходящей к случаю функцией.

Но могу предложить другой подход, более удобный. Просто у меня есть опыт использования не только Windows, но и некоторых "более других" операционных систем. И в одной из них в диалоге выбора файла был а) выпадающий список ранее использованных каталогов и б) выпадающий список заданных пользователем каталогов. Очень удобная вещь.
Для Windows, насколько я знаю, тоже есть программы, добавляющие в стандартный диалог нечто подобное.

**Comment by NS-Clone** [comment]
> Не очень понятно, чего тебе хочется.

я хочу ОДНУ ПАПКУ загрузок заданную в настройках
а не разросанные файлы по рандомным папкам
чего ТУТ понимать надо?

**Comment by zanud** [comment]
> я хочу ОДНУ ЕДИНУЮ ПАПКУ загрузок заданную в настройках

Но ведь по умолчанию это именно так и есть: все сохранения идут в каталог, указанный в "Save files to". Чем же ты недоволен?

> от туда будет читатсо только при запуске
> а это решает только часть проблемы

Настройку можно задавать через `prefs.js`, `user.js` и `.cfg`, используя разные функции (см., например,  https://support.mozilla.org/en-US/kb/customizing-firefox-using-autoconfig).

В зависимости от того, какая функция использована, значение настройки либо будет меняться динамически в процессе работы браузера, но не сохраняться при выходе (и тогда, соответственно, его записанное в файл значение заново подхватится при следующем запуске), либо вообще не будет меняться (и тогда тот же диалог выбора файла будет каждый раз открываться на одном и том же каталоге).

> тобешь нужна вообще опциональная неприкосновенность lastdir-а
> даже на время сессии

Это тебе, скорее всего, `lockPref()` надо.

**Comment by NS-Clone** [comment]
> Это тебе, скорее всего, `lockPref()` надо

это как?

**Comment by zanud** [comment]
Я ссылку на инструкцию дал. Хотя, возможно, и из `user.js` сработает.

То есть, идёшь в свой профиль Mypal, залезаешь в `prefs.js`, находишь там строку с `browser.download.lastDir`, копируешь её в буфер, создаёшь в профиле `user.js`, вставляешь в него скопированную строку и меняешь в ней `user_pref` на `lockPref`, после чего запускаешь браузер и проверяешь, что получилось.

Если повезло и при каждом сохранении тебе предлагает родин и тот же каталог - радуешься.
Если не повезло - создаёшь .cfg-файл по той инструкции.

> и все время это новый браузер без этого user.js по какойто причине

Всё правильно, браузер никогда сам это файл не создаёт. Но и не убивает, если он есть в профиле, доставшемся ему от предыдущей версии.

**Comment by NS-Clone** [comment]
user.js c lockPref("browser.download.lastDir", "X:\XXXX\XXXX");
непомогает

**Comment by zanud** [comment]
Ну, во-первых, обратные косые в строках надо удваивать.
Но да, не помогает, я тоже уже проверил.

Значит, надо cfg-файл делать, там точно работает. (Точнее, работало во времена FF 53).

**Comment by NS-Clone** [comment]
> Ну, во-первых, обратные косые в строках надо удваивать.

я скопировал из prefs.js

**Comment by zanud** [comment]
Придётся тебя огорчить: на lockPref Mypal ругается в консоль браузера:

`prefs parse error: unknown keyword`

**Comment by NS-Clone** [comment]
с cfg вроде сработало

**Comment by zanud** [comment]
> а ты написал 1-й строкой комент
> и свои значения ровно со второй строки?

Это-то я как надо сделал. Просто в той строке опетачку допустил. Заметил, исправил - заработало.

**Comment by NS-Clone** [comment]
теперь бы такое же для спeрмиума
ато там вообще клиника
